pkg/pillar/zfs/zfs.go

```diff
@@ -189,3 +193,103 @@ func GetZFSVolumeInfo(log *base.LogObject, device string) (*types.ImgInfo, error
+
+	for _, volume := range dataset.Children {
+		pathD, err := volume.Path()
+		pr, err := volume.GetProperty(libzfs.DatasetPropType)
+		if err != nil {
+			return uint32(count), fmt.Errorf("get property for dataset %s failed %v", pathD, err)
+		}
+
+		if pr.Value == "volume" {
+			count++
+		}
+	}
+
+	return uint32(count), nil
+}
+
+// getRaidTypeFromStr takes a RAID name as input and returns current RAID type
+func getRaidTypeFromStr(raidName string) info.StorageRaidType {
+	if len(raidName) == 0 {
+		return info.StorageRaidType_STORAGE_RAID_TYPE_NORAID
+	} else if strings.Contains(raidName, "raidz1") {
+		return info.StorageRaidType_STORAGE_RAID_TYPE_RAIDZ1
+	} else if strings.Contains(raidName, "raidz2") {
+		return info.StorageRaidType_STORAGE_RAID_TYPE_RAIDZ2
+	} else if strings.Contains(raidName, "raidz3") {
+		return info.StorageRaidType_STORAGE_RAID_TYPE_RAIDZ3
+	} else if strings.Contains(raidName, "mirror") {
+		return info.StorageRaidType_STORAGE_RAID_TYPE_RAID_MIRROR
+	}
+
+	return info.StorageRaidType_STORAGE_RAID_TYPE_NORAID
+}
+
+// GetZpoolRaidType takes a libzfs.VDevTree as input and returns current RAID type.
+// At the moment, while will start from the fact that for one pool, have one RAID
+func GetZpoolRaidType(vdevs libzfs.VDevTree) info.StorageRaidType {
+	for _, vdev := range vdevs.Devices {
+		if vdev.Type == libzfs.VDevTypeMirror || vdev.Type == libzfs.VDevTypeRaidz {
+			return getRaidTypeFromStr(vdev.Name)
+		}
+		break
+	}
+
+	return info.StorageRaidType_STORAGE_RAID_TYPE_NORAID
+}
+
+// GetZfsDeviceStatusFromStr takes a string with status as input and returns status
+func GetZfsDeviceStatusFromStr(statusStr string) metrics.StorageStatus {
+	if len(statusStr) == 0 {
+		return metrics.StorageStatus_STORAGE_STATUS_UNSPECIFIED
+	} else if strings.TrimSpace(statusStr) == "ONLINE" {
+		return metrics.StorageStatus_STORAGE_STATUS_ONLINE
+	} else if strings.TrimSpace(statusStr) == "DEGRADED" {
+		return metrics.StorageStatus_STORAGE_STATUS_DEGRADED
+	} else if strings.TrimSpace(statusStr) == "FAULTED" {
+		return metrics.StorageStatus_STORAGE_STATUS_FAULTED
+	} else if strings.TrimSpace(statusStr) == "OFFLINE" {
+		return metrics.StorageStatus_STORAGE_STATUS_OFFLINE
+	} else if strings.TrimSpace(statusStr) == "UNAVAIL" {
+		return metrics.StorageStatus_STORAGE_STATUS_UNAVAIL
+	} else if strings.TrimSpace(statusStr) == "REMOVED" {
+		return metrics.StorageStatus_STORAGE_STATUS_REMOVED
+	} else if strings.TrimSpace(statusStr) == "SUSPENDED" {
+		return metrics.StorageStatus_STORAGE_STATUS_SUSPENDED
+	}
+
+	return metrics.StorageStatus_STORAGE_STATUS_UNSPECIFIED
+}
+
+// GetZfsDiskAndStatus takes a libzfs.VDevTree as input and returns
+// *metrics.StorageDiskInfo.
+func GetZfsDiskAndStatus(disk libzfs.VDevTree) (*metrics.StorageDiskInfo, error) {
+	if disk.Type != libzfs.VDevTypeDisk {
+		return nil, fmt.Errorf("%s is not a disk", disk.Name)
+	}
+
+	rDiskStatus := new(metrics.StorageDiskInfo)
+	rDiskStatus.DiskName = disk.Name
```

Comment:
For some reason I can see `/dev/sdc1` instead of `/dev/sdc` or just `sdc`. But it is just observation, not a blocker.

Reply:
Well, do not cut off 1 if zfs divides disks into partitions.